### ISSUE TITLE
No warning for comparison of integers of different signs

### ISSUE THREAD

**Issue by chengniansun** [llvm/llvm-project#18878]
|  |  |
| --- | --- |
| Bugzilla Link | [18504](https://llvm.org/bz18504) |
| Version | trunk |
| OS | Linux |
| CC | @zygoloid,@zhendongsu |

## Extended Description 
Clang does not emit a warning for the comparison of integers of different signs.

I have the following program: 

```c
int unreported(unsigned a, int *b) {
        return a > (~(95 != *b));
}
```

And compile it with the following command, but Clang emits no warning for the expression "a > (~(95 != *b))"
```console
$clang -Wall -Wextra -std=c11 -pedantic -c s.c
$clang --version
clang version 3.5 (trunk 198918)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```

Differently, GCC emits a warning with the following command
```console
gcc -Wall -Wextra -std=c11 -pedantic -c s.c

s.c: In function ‘unreported’:
s.c:7:11: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
  return a > (~(95 != *b));
           ^
```

**Comment by AshleyRoll** [comment]
I have made an initial attempt at fixing this issue as a first time contributor.

https://reviews.llvm.org/D126960

**Comment by AvoidMe** [comment]
Hi!

I see that there was no progress on this one for some time.
Can I take @AshleyRoll patch rebase & fix comments?

**Comment by AshleyRoll** [comment]
@AvoidMe feel free to take over the changes if you like. Unfortunately I've been too busy to get back to this like I had hoped.

**Comment by YutongZhuu** [comment]
Hi, I would like to be assigned to this issue

**Comment by YutongZhuu** [comment]
Sorry, I don't understand what the issue is. The code is comparing a int with a boolean, why should we throw a warning about comparing signed and unsigned int? If we change the test code to:
```c
int unreported(unsigned a, int *b) {
    return a  > *b;
}
```
I get
```
❯ clang -Wall -Wextra -Wsign-compare -std=c11 -pedantic -c test.c
test.c:2:10: warning: comparison of integers of different signs: 'unsigned int' and 'int' [-Wsign-compare]
    2 | return a > *b;
      |        ~ ^ ~~
1 warning generated.
```
I don't understand why is the issue still opened?

**Comment by shafik** [comment]
So the result of the `~` in the original problem will complement a bool which will always be negative and then promote it to `unsigned` which will be some really large unsigned value. This is problematic is likely not intended.

A comment earlier pointed out this code:

https://github.com/llvm/llvm-project/blob/da2b415b19d178d6b250bb6468ed4a207860adf7/clang/lib/Sema/SemaChecking.cpp#L10071-L10074

so it will take the subexpr, which being a relation expression will result in the range being treated as `bool` which is strictly bounded `[0,1]` which is always in the range of `unsigned` and thus well behaved. 

So we need to special case the switch to cover in this specific case `UO_Not` but as pointed out it is wrong for at least `UO_PreInc`, `UO_PreDec`, `UO_Minus` as well.

**Comment by YutongZhuu** [comment]
> So the result of the `~` in the original problem will complement a bool which will always be negative and then promote it to `unsigned` which will be some really large unsigned value. This is problematic is likely not intended.

There would be a warning generated if one is trying to negate a bool. 
```c
❯ clang -Wall -Wextra -Wsign-compare -std=c11 -pedantic -c test.c
test.c:2:13: warning: bitwise negation of a boolean expression; did you mean logical negation? [-Wbool-operation]
    2 | return a > (~(95 != *b));
      |             ^~~~~~~~~~~
      |             !
1 warning generated.
```

**Comment by shafik** [comment]
So the `Wbool-operation` diagnostic is indeed valid. 

After the complement of a *bool* we are left with an *int* because of *integral promotions*, which will have the value `-1` or `-2` which are both outside the range of values for *unsigned*. Then the value will be converted due to the usual arithmetic conversion being applied to the operands of relational expression, to *unsigned* which will result in some very large unsigned value.

So the `Wsign-compare` diagnostic is for the mismatch between the RHS and LHS of the relational expression before the usual arithmetic conversions are applied. 

OTOH with:

```c
return a > (95 != *b);
```

the result of the RHS is a boolean value, which while will be converted to *int* is bounded between `[0,1]` and which are within the value range of *unsigned* and so no concerns there.

**Comment by YutongZhuu** [comment]
> So the `Wbool-operation` diagnostic is indeed valid.
> 
> After the complement of a _bool_ we are left with an _int_ because of _integral promotions_, which will have the value `-1` or `-2` which are both outside the range of values for _unsigned_. Then the value will be converted due to the usual arithmetic conversion being applied to the operands of relational expression, to _unsigned_ which will result in some very large unsigned value.
> 
> So the `Wsign-compare` diagnostic is for the mismatch between the RHS and LHS of the relational expression before the usual arithmetic conversions are applied.
> 
> OTOH with:
> 
> return a > (95 != *b);
> the result of the RHS is a boolean value, which while will be converted to _int_ is bounded between `[0,1]` and which are within the value range of _unsigned_ and so no concerns there.

That makes a lot of sense, so basically the problem is a bool gets converted to a int, and after that we should warn the user for comparing int and unsigned int. Makes sense, I was not reading the original post correctly, sorry for wasting your precious time. Btw, I would like to be assigned to this issue if possible, thanks!

**Comment by shafik** [comment]
> That makes a lot of sense, so basically the problem is a bool gets converted to a int, and after that we should warn the user for comparing int and unsigned int. Makes sense, I was not reading the original post correctly, sorry for wasting your precious time. Btw, I would like to be assigned to this issue if possible, thanks!

Done, it is not a problem. The point is helping new folks is that they can hopefully over the long-term become continuous contributors. We can't grow w/o helping new folks get started.

**Comment by YutongZhuu** [comment]
@shafik Hi, I submitted a PR regarding the issue with the ``~`` operator. However, if I apply the same fix to ``-`` operator, the following test fails (clang/test/Sema/unary-minus-integer-impcast.c): 

```c++
void test(void) {
  unsigned int a = 1;
#endif
  long c2 = -a; // expected-warning@{{implicit conversion changes signedness: 'unsigned int' to 'long'}}
}
```

This warning will only be threw if the right hand-side is non-negative. By making the fix, any expression with ``-`` will not be marked as Non-Negative, which I think makes more sense, should I go ahead and find another way to throw this warning(``implicit conversion changes signedness: 'unsigned int' to 'long'``)?  I would like to know your opinion before putting any effort into this.
